### DISCUSSION TITLE
Team and User Layered configuration using `chezmoi`

### DISCUSSION THREAD

**Discussion by Lockszmith-GH** [twpayne/chezmoi#2574] [Show and tell]
`chezmoi` is really versatile, I guess you already know this, but I thought I'd share our particular use-case.

# Who
We are a team of remote workers, almost all of us work on Linux machines (Pop!_OS), we have a few mac OS laptops and one or two Windows machine.
Currently `chezmoi` covers the Linux realm, but we might expand this in the future.
We all work with GitLab SaaS under a Premium subscription, and all of repos are internal. (They are shared with the team, but not with the world).

# What
I've developed a shared environment for the Linux users to use.
This environment does a lot - adds needed repos to apt on the local system, provides 'easy install' scripts, keep external resources up to date, customizes the work environment (fonts, prompts, auto-completion, history, git configuration, etc...), add support scripts to allow IT-Ops to quickly connect remotely and assist.
So, in short: A whole bunch of $h!t

I've been working on a layered approach to `chezmoi`, one that would allow each team member to manage their own personal `chezmoi` repo, but this personal repo is tied to a shared team configuration which provides shared functionality and configuration.

# Why
By layering, if a team member wants to customize something on their own, they have the ability, but still get streamlined updates that are available to the entire team.
This makes `chezmoi` not only the safe-keeper of personal state, but a convenient delivery channel of updates to the **base**.

# How
##  Structure
Each user gets a fork of a template repo when they start. That repo has the following components:
* `.chezmoi.toml.tmpl`
  * This contains a read-only repo specific authentication token (`dotfiles_git_token`) that is used to pull the shared repo (aka **base**)
  * The value is initialized by prompting the user to enter it on first `init` - so this _secret_ is never stored in the repo.
* `.chezmoitexternals.toml`
   * Points to the `0base` _dotfiles_ repo which contains all of the shared functionality.
   * Since all of our repos are private, including the **base** repo requires credentials. The URL includes a template string pulling the `dotfiles_git_token` value.
* `.chezmoiscripts` directory with:
  * `run_once_after_10_init_base_chezmoi`
     * executes `init` on the **base** chezmoi after it was pulled via _externals_.
  * `run_after_apply_base_chezmoi.sh`
     * executed on every `apply` - applying any changes pulled from the _external_ **base**.

# Usage
The `0base` chezmoi-repo contains environment variables, scripts and aliases that provide different shorthand for calling `chezmoi`:
* `cz` - the user's `chezmoi` context - pointing to the default `chezmoi` dir:
  > `~/.local/chezmoi/`
* `czb` - the **base** chezmoi context - pointing to a **base** `chezmoi` dir:
  > `~/.local/chezmoi.base/`

## Context switching
For some 'more complex' calls, the context concept is designated using the `CONTEXT` environment variable.
For example: `cz-init` is a script used to call `chezmoi init` with the proper `--script-path` parameter, so that values are written back into the same source config file it is called with.
Running `cz-init` does this on the _default_ (user) chezmoi repo.
Running `CONTEXT=czb cz-init` however, will run it in the **base** context.

# But wait, there's is more...
There is more complexity the shared repo provides, like [separating externals](https://github.com/twpayne/chezmoi/discussions/2375#discussioncomment-3735286) from the **base** configuration, which allows quicker responses from `chezmoi status` and makes maintaining and tracking changes faster (less accurate, but faster) from the **base** configuration, which allows quicker responses from chezmoi status

# Code? (sometime?)
~~I'm almost ready to share the code, but not quite yet, as it's a bit messy (also need to have one of my colleagues validate I will not be exposing any secrets/password), but if you're interested, just :eyes: this post, and I'll update below when I'm ready.~~
I've never got to clean it up and publish - it is still in use. Ask questions and I'll do my best to answer.

# Visual

```mermaid %%{ init: { 'theme': 'dark' }}%%
flowchart TD
    h{{home}}
    subgraph user ["~/.local/share/chezmoi"]
        cz[[chezmoi<br>local-state]]
    end

    subgraph base ["~/.local/share/chezmoi.base"]
        czbr([".chezmoiroot"])
        czb[[base-chezmoi<br>local-state]]

        subgraph ext ["~/.local/share/chezmoi.base/_externals"]
            czx[[externals-chezmoi<br>local-state]]
        end        
    end

    subgraph git ["Remote git repos (GitHub)"]
        czg[("dotfiles.< USER NAME >.git")]
        czbg[("dotfiles.0base.git")]
    end
    
    h       ===> |cz commands     | cz
    user   <--> |git             | czg
    h      ---- |czg commands    | czg
    czbr   -.-> |_home.base      | czb
    h      ======> |czb commands    | czb
    cz     -..- |externals       | base
    czb     -.- |externals       | ext
    base <----> |git             | czbg
```

# Praise
I couldn't have done this without the support of Tom Payne the author, his support for this amazing tool is amazing on it's own level.
Thank you! (I didn't tag Tom as I assume he has enough noise as it is)

# What about you?
I'm really interested to learn if others have developed other interesting patterns of usage with `chezmoi`?
Did you ever face similar needs?
As we will venture into multi-platform coverage, I wonder about best approaches.

**Comment by Lockszmith-GH** [comment]
Updates and backlinks

**Reply by Lockszmith-GH** [reply]
* Mentioned in #2673 [🔗comment](https://github.com/twpayne/chezmoi/discussions/2673#discussioncomment-4985961)
* Mentioned in #2631 [🔗comment](https://github.com/twpayne/chezmoi/discussions/2631#discussioncomment-4722259)
* Added Visual section with a [mermaid.js](https://mermaid.js.org/syntax/flowchart.html) diagram.

**Reply by Lockszmith-GH** [reply]
* Mentioned in #3083 [🔗comment](https://github.com/twpayne/chezmoi/discussions/3083#discussioncomment-6405609)

**Reply by Lockszmith-GH** [reply]
* Mentioned in #3460 [🔗comment](https://github.com/twpayne/chezmoi/discussions/3460#discussioncomment-8042242)

**Reply by Lockszmith-GH** [reply]
* Mentioned in #3659 [🔗comment](https://github.com/twpayne/chezmoi/discussions/3659#discussioncomment-8897668)

**Reply by Lockszmith-GH** [reply]
* Mentiond in #3768 [🔗comment](https://github.com/twpayne/chezmoi/discussions/3768#discussioncomment-9476722)

**Reply by Lockszmith-GH** [reply]
* Mentioned in #3773

**Reply by Lockszmith-GH** [reply]
* Updated Code section, from soon to 'sometime' - as life has been [OBE](https://www.merriam-webster.com/dictionary/overtaken%20by%20events) and I don't know when I'll get to it.

**Comment by Lockszmith-GH** [comment]
Placeholder for future 2

**Comment by lnyousif** [comment]
This is brilliant approach, is there any repo link that I am missing for this ?

**Reply by Lockszmith-GH** [reply]
I have never got to publish it, as I hadn't really sanitized the internal repo.
I can try and tailor something and share.
If you have questions, I would be happy to answer.

**Reply by lnyousif** [reply]
if you want to take down all the secrets and put empty files, I can help with prepping an example that you can share publicly later. Just a thought

**Reply by Lockszmith-GH** [reply]
Been busy with some stuff, I'll try and get to it, but can't commit to anything.

**Comment by PeterCardenas** [comment]
@Lockszmith-GH what are you setting for the index in the externals file? I'm trying to set an empty path so that all external files are merged with the base one, but getting an `empty path` error

```toml
[""]
type = "git-repo"
url = "<url to repo>"
```

**Reply by Lockszmith-GH** [reply]
Not sure I'm understanding.

Can you explain what is the end state you want to achieve?

**Reply by PeterCardenas** [reply]
i have a private dot files that i'm trying to make public by separating out the work specific stuff from the agnostic stuff. some files should only exist for work so im moving these to a separate repository. these files only have a common directory of ~ so i want to include all the files from the private repository and place them as is. for example, let's say i have a file ~/.work and a directory -/.config/work i need to take from the private repo. I'd prefer not to expose that the specific files exist but rather just include them all from the private repo

**Reply by Lockszmith-GH** [reply]
I see, and you are trying to pull those files using externals?
You might be able to do that with sub-dirs, but as full directories - you will not be successful, at least not with chezmoi.

What I did in this article is something else.

I maintain 2 separate chemozi instances.
And when initializing one, it initilizes the other as well via scripts.

then I manage these 2 contexts side-by-side.

If that's what you want to do, I'd be happy to explain more.

**Reply by PeterCardenas** [reply]
yea two source states is what i was originally looking for which is how i found this discussion. from the diagram it seemed like you were using externals for this which is why i mentioned using externals

**Reply by Lockszmith-GH** [reply]
no, I probably confused everything there even more.
I call each of these chezmoi instances 'CONTEXT' or chezmoi context.

The externals was a separate context of it's own - because I was pulling so many externals, that bogged down and caused slowness - today I do something different, I just wrap the externals file with an {{- if -}} block, and enable it when it's time to check externals.

Anyway, the way I use 2 separate chezmoi contexts, was by having separate command (I describe these are `cz` and `czb`)

It looks like this:
`cz` controls ~/.local/share/chezmoi - the default chezmoi instance.  
`czb` controls ~/.local/share/chezmoi.base - the 'base' chezmoi instance (the shared instance).

`cz` is just plain `chezmoi` (an alias or function will do the trick)

`czb` is something like:

```bash
chezmoi --source "${USER_HOME:-$HOME}/.local/share/chezmoi.base" --config ${USER_HOME:-$HOME}/.config/chezmoi/_base/chezmoi.toml "${@}"
```

I hope this helps understand what's going on up there.

**Reply by PeterCardenas** [reply]
nice! --source sounds like exactly what i was looking for

**Reply by PeterCardenas** [reply]
just for my knowledge externals mean every time you apply it fetches the externals?

**Reply by LEI** [reply]
@PeterCardenas `refreshPeriod` is set to 0 by default meaning "never re-download", you can find more details following the link to the reference manual here: https://www.chezmoi.io/user-guide/include-files-from-elsewhere/

**Reply by Lockszmith-GH** [reply]
However, it's not just refreshPeriod, the state of externals is read every time and compared to the active state, regardless of online state.
So while it's not reaching online, it is reading the local cache - if that cache is large, and disk is slow - it takes time.

For more details, see #2375
